File: src/Ninja.cpp
#include "KwestKingdom.h"
#include "Ninja.h"
#include "resources.h"
#include "Sword.h"
#include "utilities.h"
#include "World.h"


typedef enum {
  NINJA_STAND_STATE,
  NINJA_ATTACK_STATE,
  NINJA_MOVE_STATE,
  NINJA_DASH_STATE
} NINJA_STATE;


Ninja::Ninja() {
  
  x = 0;
  y = 0;
  
  team = ENEMY_TEAM;
  
  sword = new Sword();
  sword->setSpeed(speed);
  sword->toHoldState();
  
  init_anim(&stand_anim, ON, 6);
  add_frame(&stand_anim, IMG("ninja_1.bmp"));
  add_frame(&stand_anim, IMG("ninja_2.bmp"));
  add_frame(&stand_anim, IMG("ninja_3.bmp"));
  add_frame(&stand_anim, IMG("ninja_2.bmp"));
  
  copy_anim(&dash_anim, &stand_anim);
  dash_anim.speed = 24;
  
  init_anim(&attack_anim, OFF, 12);
  add_frame(&attack_anim, IMG("ninja_1.bmp"));
  add_frame(&attack_anim, IMG("ninja_2.bmp"));
  add_frame(&attack_anim, IMG("ninja_2.bmp"));
  add_frame(&attack_anim, IMG("ninja_3.bmp"));
  add_frame(&attack_anim, IMG("ninja_3.bmp"));
  add_frame(&attack_anim, IMG("ninja_2.bmp"));
  add_frame(&attack_anim, IMG("ninja_2.bmp"));
  
  anim = &stand_anim;
  state = NINJA_STAND_STATE;
  this->setSpeed(getWalkSpeed());
  this->wait();
}


Ninja::~Ninja() {
  delete sword;
}


void Ninja::update() {
  
  int dir;
  int toX;
  int toY;
  Character *target;
  
  Enemy::update();
  sword->update();
  
  if (this->waiting()) {
    return;
  }
  
  if (health == 0) {
    return;
  }
  
  target = world->getTarget();
  
  switch (state) {
  
  case NINJA_STAND_STATE:
    
    // If the target has a walking distance of one...
    if (abs(x - target->getX()) + abs(y - target->getY()) == 1) {
      
      state = NINJA_ATTACK_STATE;
      anim = &attack_anim;
      reset_anim(anim);
      // Change the state of the sword.
      if (x == target->getX() && y == target->getY() + 1) { // Up
        sword->toAttackUpState();
      } else if (x == target->getX() && y == target->getY() - 1) { // Down
        sword->toAttackDownState();
      } else if (x == target->getX() + 1 && y == target->getY()) { // Left
        sword->toAttackLeftState();
      } else if (x == target->getX() - 1 && y == target->getY()) { // Right
        sword->toAttackRightState();
      }
      
    } else if (x == target->getX()) {
      
      if (y > target->getY()) { // Hero is directly up.
        while (y - 1 != target->getY() && world->isWalkable(x, y - 1) && !world->isInhabited(x, y - 1)) {
          this->moveY(y - 1);
        }
      } else { // Hero is directly down.
        while (y + 1 != target->getY() && world->isWalkable(x, y + 1) && !world->isInhabited(x, y + 1)) {
          this->moveY(y + 1);
        }
      }
      
      state = NINJA_DASH_STATE;
      this->setSpeed(getWalkSpeed() + (getWalkSpeed() / 5));
      anim = &dash_anim;
      reset_anim(anim);
      
    } else if (y == target->getY()) {
      
      if (x > target->getX()) { // Hero is directly left.
        while (x - 1 != target->getX() && world->isWalkable(x - 1, y) && !world->isInhabited(x - 1, y)) {
          this->moveX(x - 1);
        }
      } else { // Hero is directly right.
        while (x + 1 != target->getX() && world->isWalkable(x + 1, y) && !world->isInhabited(x + 1, y)) {
          this->moveX(x + 1);
        }
      }
      
      state = NINJA_DASH_STATE;
      this->setSpeed(getWalkSpeed() + (getWalkSpeed() / 5));
      anim = &dash_anim;
      reset_anim(anim);
      
    } else {
      
      // Wander aimlessly
      dir = random_number(UP, /*DOWN, LEFT,*/ RIGHT);
      
      toX = x;
      toY = y;
      
      if (dir == UP) {
        toY--;
      } else if (dir == DOWN) {
        toY++;
      } else if (dir == RIGHT) {
        toX++;
      } else if (dir == LEFT) {
        toX--;
      }
      
      if (world->isWalkable(toX, toY) && !world->isInhabited(toX, toY)) {
        this->moveX(toX);
        this->moveY(toY);
        state = NINJA_MOVE_STATE;
      }
      
      this->wait();
      
    }
    
    // Bound him so he doesn't wander right out of the screen!
    this->bound(1, ROWS - 2, 1, COLS - 2);
    
    break;
    
  case NINJA_MOVE_STATE:
    if (!this->moving()) {
      state = NINJA_STAND_STATE;
    }
    break;
    
  case NINJA_DASH_STATE:
    if (!this->moving()) {
      this->setSpeed(getWalkSpeed());
      // If the target has a walking distance of one...
      if (abs(x - target->getX()) + abs(y - target->getY()) == 1) {
        state = NINJA_ATTACK_STATE;
        anim = &attack_anim;
        reset_anim(anim);
        // Change the state of the sword.
        if (x == target->getX() && y == target->getY() + 1) { // Up
          sword->toAttackUpState();
        } else if (x == target->getX() && y == target->getY() - 1) { // Down
          sword->toAttackDownState();
        } else if (x == target->getX() + 1 && y == target->getY()) { // Left
          sword->toAttackLeftState();
        } else if (x == target->getX() - 1 && y == target->getY()) { // Right
          sword->toAttackRightState();
        }
      } else {
        state = NINJA_STAND_STATE;
        anim = &stand_anim;
        reset_anim(anim);
        this->wait();
      }
    }
    break;
    
  case NINJA_ATTACK_STATE:
    if (anim->done) {
      world->attackFromTeam(team, target->getX(), target->getY());
      state = NINJA_STAND_STATE;
      anim = &stand_anim;
      reset_anim(anim);
      sword->toHoldState();
      this->wait();
    }
	break;
	
  }
  

  
}


void Ninja::draw(IMAGE * canvas) {
  Enemy::draw(canvas);
  sword->draw(canvas);

}


void Ninja::setX(int newX) {
  Enemy::setX(newX);
  sword->setX(newX);

}


void Ninja::setY(int newY) {
  Enemy::setY(newY);
  sword->setY(newY);

}


void Ninja::moveX(int newX) {
  Enemy::moveX(newX);
  sword->moveX(newX);

}


void Ninja::moveY(int newY) {
  Enemy::moveY(newY);
  sword->moveY(newY);

}


void Ninja::setSpeed(int theSpeed) {
  speed = theSpeed;
  sword->setSpeed(speed);

}

File: src/Ninja.m
#import "Ninja.h"


typedef enum {
  NINJA_STAND_STATE,
  NINJA_ATTACK_STATE,
  NINJA_MOVE_STATE,
  NINJA_DASH_STATE
} NINJA_STATE;


@implementation Ninja


- init {
  
  self = [super init];
  
  if (self) {
    
    x = 0;
    y = 0;
    
    [self setSpeed: getWalkSpeed()];
    team = ENEMY_TEAM;
    
    sword = [[Sword alloc] init];
    [sword setSpeed: speed];
    [sword toHoldState];
    
    standAnimation = [[Animation alloc] init];
    [standAnimation addFrame: getImage(IMAGES_NINJA_1)];
    [standAnimation addFrame: getImage(IMAGES_NINJA_2)];
    [standAnimation addFrame: getImage(IMAGES_NINJA_3)];
    [standAnimation addFrame: getImage(IMAGES_NINJA_2)];
    [standAnimation setLoop: YES];
    [standAnimation setSpeed: 6];
    
    dashAnimation = [standAnimation copy];
    [dashAnimation setSpeed: 24];
    
    attackAnimation = [[Animation alloc] init];
    [attackAnimation addFrame: getImage(IMAGES_NINJA_1)];
    [attackAnimation addFrame: getImage(IMAGES_NINJA_2)];
    [attackAnimation addFrame: getImage(IMAGES_NINJA_2)];
    [attackAnimation addFrame: getImage(IMAGES_NINJA_3)];
    [attackAnimation addFrame: getImage(IMAGES_NINJA_3)];
    [attackAnimation addFrame: getImage(IMAGES_NINJA_2)];
    [attackAnimation addFrame: getImage(IMAGES_NINJA_2)];
    [attackAnimation setLoop: NO];
    [attackAnimation setSpeed: 12];
    
    animation = standAnimation;
    state = NINJA_STAND_STATE;
    [self wait];
    
  }
  
  return self;
  
}


- (void) dealloc {
  [sword release];
  [standAnimation release];
  [dashAnimation release];
  [attackAnimation release];
  [super dealloc];
}


- update {
  
  int dir;
  int toX;
  int toY;
  id<Positionable> target;
  
  [super update];
  [sword update];
  
  if ([self waiting]) {
    return self;
  }
  
  if (health == 0) {
    return self;
  }
  
  target = [world getTarget];
  
  switch (state) {
  
  case NINJA_STAND_STATE:
    
    // If the target has a walking distance of one...
    if (abs(x - [target getX]) + abs(y - [target getY]) == 1) {
      
      state = NINJA_ATTACK_STATE;
      animation = attackAnimation;
      [animation reset];
      // Change the state of the sword.
      if (x == [target getX] && y == [target getY] + 1) { // Up
        [sword toAttackUpState];
      } else if (x == [target getX] && y == [target getY] - 1) { // Down
        [sword toAttackDownState];
      } else if (x == [target getX] + 1 && y == [target getY]) { // Left
        [sword toAttackLeftState];
      } else if (x == [target getX] - 1 && y == [target getY]) { // Right
        [sword toAttackRightState];
      }
      
    } else if (x == [target getX]) {
      
      if (y > [target getY]) { // Hero is directly up.
        while (y - 1 != [target getY] && [world isWalkableAtX: x andY: y - 1] && ![world isInhabitedAtX: x andY: y - 1]) {
          [self moveY: y - 1];
        }
      } else { // Hero is directly down.
        while (y + 1 != [target getY] && [world isWalkableAtX: x andY: y + 1] && ![world isInhabitedAtX: x andY: y + 1]) {
          [self moveY: y + 1];
        }
      }
      
      state = NINJA_DASH_STATE;
      [self setSpeed: getWalkSpeed() + (getWalkSpeed() / 5)];
      animation = dashAnimation;
      [animation reset];
      
    } else if (y == [target getY]) {
      
      if (x > [target getX]) { // Hero is directly left.
        while (x - 1 != [target getX] && [world isWalkableAtX: x - 1 andY: y] && ![world isInhabitedAtX: x - 1 andY: y]) {
          [self moveX: x - 1];
        }
      } else { // Hero is directly right.
        while (x + 1 != [target getX] && [world isWalkableAtX: x + 1 andY: y] && ![world isInhabitedAtX: x + 1 andY: y]) {
          [self moveX: x + 1];
        }
      }
      
      state = NINJA_DASH_STATE;
      [self setSpeed: getWalkSpeed() + (getWalkSpeed() / 5)];
      animation = dashAnimation;
      [animation reset];
      
    } else {
      
      // Wander aimlessly
      dir = random_number(UP, /*DOWN, LEFT,*/ RIGHT);
      
      toX = x;
      toY = y;
      
      if (dir == UP) {
        toY--;
      } else if (dir == DOWN) {
        toY++;
      } else if (dir == RIGHT) {
        toX++;
      } else if (dir == LEFT) {
        toX--;
      }
      
      if ([world isWalkableAtX: toX andY: toY] && ![world isInhabitedAtX: toX andY: toY]) {
        [self moveX: toX];
        [self moveY: toY];
        state = NINJA_MOVE_STATE;
      }
      
      [self wait];
      
    }
    
    // Bound him so he doesn't wander right out of the screen!
    [self boundAtTop: 1 andBottom: ROWS - 2 andLeft: 1 andRight: COLS - 2];
    
    break;
    
  case NINJA_MOVE_STATE:
    if (![self moving]) {
      state = NINJA_STAND_STATE;
    }
    break;
    
  case NINJA_DASH_STATE:
    if (![self moving]) {
      [self setSpeed: getWalkSpeed()];
      // If the target has a walking distance of one...
      if (abs(x - [target getX]) + abs(y - [target getY]) == 1) {
        state = NINJA_ATTACK_STATE;
        animation = attackAnimation;
        [animation reset];
        // Change the state of the sword.
        if (x == [target getX] && y == [target getY] + 1) { // Up
          [sword toAttackUpState];
        } else if (x == [target getX] && y == [target getY] - 1) { // Down
          [sword toAttackDownState];
        } else if (x == [target getX] + 1 && y == [target getY]) { // Left
          [sword toAttackLeftState];
        } else if (x == [target getX] - 1 && y == [target getY]) { // Right
          [sword toAttackRightState];
        }
      } else {
        state = NINJA_STAND_STATE;
        animation = standAnimation;
        [animation reset];
        [self wait];
      }
    }
    break;
    
  case NINJA_ATTACK_STATE:
    if ([animation finished]) {
      [world attackFromTeam: team atX: [target getX] andY: [target getY]];
      state = NINJA_STAND_STATE;
      animation = standAnimation;
      [animation reset];
      [sword toHoldState];
      [self wait];
    }
	break;
	
  }
  
  return self;
  
}


- draw: (BITMAP *) buffer {
  [super draw: buffer];
  [sword draw: buffer];
  return self;
}


- setX: (int) newX {
  [super setX: newX];
  [sword setX: newX];
  return self;
}


- setY: (int) newY {
  [super setY: newY];
  [sword setY: newY];
  return self;
}


- moveX: (int) newX {
  [super moveX: newX];
  [sword moveX: newX];
  return self;
}


- moveY: (int) newY {
  [super moveY: newY];
  [sword moveY: newY];
  return self;
}


- setSpeed: (int) theSpeed {
  speed = theSpeed;
  [sword setSpeed: speed];
  return self;
}


@end
